scripts/gen-exprs.py

```diff
@@ -0,0 +1,943 @@
+    """A binary opcode."""
+
+
+class Child(Field):
+    """A child expression."""
+
+    type_name = 'Expression*'
+
+
+class ChildList(Field):
+    """A list of child expressions."""
+
+    type_name = 'ExpressionList'
+
+    allocator = True
+
+
+class AtomicRMWOp(Field):
+    """An atomic RMW opcode."""
+
+
+class SIMDExtractOp(Field):
+    """A SIMD Extract opcode."""
+
+
+class SIMDReplaceOp(Field):
+    """A SIMD Replace opcode."""
+
+
+class SIMDShuffleMask(Field):
+    """An SIMD shuffle mask."""
+
+    type_name = 'std::array<uint8_t, 16>'
+
+
+class SIMDTernaryOp(Field):
+    """A SIMD ternary opcode."""
+
+
+class SIMDShiftOp(Field):
+    """A SIMD shift opcode."""
+
+
+class SIMDLoadOp(Field):
+    """A SIMD load opcode."""
+
+
+class Method:
+    """A function method on an expression class."""
+
+    def __init__(self, paramses, result, const=False):
+        self.paramses = paramses
+        if type(self.paramses) is str:
+            self.paramses = [self.paramses]
+        self.result = result
+        self.const = const
+
+
+class Expression:
+    """Core class from which all expressions inherit."""
+
+    # Optional content to place in the constructor body.
+    __constructor_body__ = ''
+
+    # If no finalize() method is defined, a default one is emitted.
+    default_finalize = Method('', 'void')
+
+    @classmethod
+    def get_fields(self):
+        """Get a map of field name => field."""
+        fields = {}
+        for key, value in self.__dict__.items():
+            if is_subclass_of(value.__class__, Field):
+                fields[key] = value
+        return fields
+
+    @classmethod
+    def get_methods(self):
+        """Get a map of method name => method."""
+        methods = {}
+        for key, value in self.__dict__.items():
+            if is_a(value, Method):
+                methods[key] = value
+        return methods
+
+
+###################################
+# Specific expression definitions
+###################################
+
+
+class Nop(Expression):
+    pass
+
+
+class Block(Expression):
+    name = ScopeNameDef()
+    list = ChildList()
+
+    """
+    finalize has three overloads:
+
+    void ();
+
+    set the type purely based on its contents. this scans the block, so it is
+    not fast.
+
+    void (Type);
+
+    sets the type given you know its type, which is the case when parsing
+    s-expression or binary, as explicit types are given. the only additional
+    work this does is to set the type to unreachable in the cases that is
+    needed (which may require scanning the block)
+
+    void (Type type_, bool hasBreak);
+
+    set the type given you know its type, and you know if there is a break to
+    this block. this avoids the need to scan the contents of the block in the
+    case that it might be unreachable, so it is recommended if you already know
+    the type and breakability anyhow.
+    """
+    finalize = Method(('', 'Type type_', 'Type type_, bool hasBreak'), 'void')
```

Comment:
It might be more readable to just take the full C++ method declarations as strings and inject them unmodified into the output.

Reply:
Maybe you're right. I pushed a commit with that now.

It is definitely simpler, so it's probably better. But it is less declarative, so if we need to do any analysis on them, that would be harder. But maybe we won't. And we can always change this later pretty easily.